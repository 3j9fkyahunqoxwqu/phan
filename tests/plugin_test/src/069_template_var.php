<?php

/**
 * @template T
 * @var T $x
 * @param T $y
 * @return T
 */
function test_var_creation($y) {
    // Some unanalyzable code that creates $x
    $a = substr('xtra', 0, 1);
    $$a = $y;  // TODO: Should detect the usage of $a
    echo strlen($x);  // should warn about misusing T
    return [$x];  // should warn about returning [T] instead of T
}
test_var_creation(new stdClass());
