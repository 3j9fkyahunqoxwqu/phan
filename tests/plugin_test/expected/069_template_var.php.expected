src/069_template_var.php:11 PhanUnusedVariable Unused definition of variable $a
src/069_template_var.php:12 PhanPluginDollarDollar $$ Variables are not allowed.
src/069_template_var.php:13 PhanTypeMismatchArgumentInternal Argument 1 (string) is T but \strlen() takes string
src/069_template_var.php:14 PhanTypeMismatchReturn Returning type array{0:T} but test_var_creation() is declared to return T
